<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="Common reference genomes hosted on AWS S3">
  <meta name="author" content="Philip Ewels">

  <title>AWS iGenomes</title>

  <!-- Bootstrap core CSS -->
  <link href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-alpha.6/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-rwoIResjU2yc3z8GV/NPeZWAv56rSmLldC3R/AZzGRnGxQQKnKkoFVhFQhNUwEyJ" crossorigin="anonymous">

  <style type="text/css">
    .mainbody {
      padding: 50px 15px;
    }
    .subtitle {
      font-weight: 200;
    }
    pre {
      background-color: #ededed;
      border-radius: 5px;
      padding: 20px;
    }
    code {
      color: inherit;
    }
    label {
      font-weight: bold;
    }
    button, .local_code {
      cursor: pointer;
    }
    hr {
      margin: 30px 0;
    }
    .prewrap {
      position: relative;
      display: -webkit-inline-box;
      display: -webkit-inline-flex;
      display: -ms-inline-flexbox;
      display: inline-flex;
      vertical-align: middle;
      width: 100%;
      margin-bottom: 10px;
    }
    .prewrap > .btn, .prewrap > pre {
      position: relative;
      -webkit-box-flex: 0;
      -webkit-flex: 0 1 auto;
      -ms-flex: 0 1 auto;
      flex: 0 1 auto;
      margin: 0;
    }
    .prewrap > .btn {
      border-bottom-right-radius: 0;
      border-top-right-radius: 0;
    }
    .prewrap > pre {
      border-bottom-left-radius: 0;
      border-top-left-radius: 0;
      width: 100%;
    }
    #manifest_heading a {
      color: #666;
    }
    .footer {
      background-color: #ededed;
      color: #999;
      padding: 20px 0;
      font-size: 0.8em;
    }
    .footer a {
      color: #999;
      text-decoration: underline;
    }
    .footer img {
      height: 60px;
    }
    .footer .sll-logo {
      margin-top: 20px;
    }
    @media only screen and (max-width: 500px) {
      .mainbody {
        padding-top: 10px;
      }
      h1.display-1 {
        font-size: 2.5rem;
      }
      h2 {
        font-size: 1.2rem;
      }
      pre {
        padding: 10px;
      }
    }
  </style>

</head>

<body>

  <div class="container mainbody">

    <img title="AWS Programs for Research and Education" src="images/AWS_logo.png" class="hidden-lg-down float-lg-right">
    <h1 class="display-1">AWS iGenomes</h1>
    <h2 class="subtitle text-muted mb-5">Common reference genomes hosted on AWS S3</h2>

    <p>This resource hosts commonly used bioinformatics reference genomes with the help
      of a grant from <a href="https://aws.amazon.com/grants/">AWS Programs for Research and Education</a>.</p>
    <p>In order to get the references, you need to sync the files from S3 to your EC2 environment.
      This web page contains two tools to help you with that - a <a href="#sync-command-builder">command builder</a>
      and a <a href="#command-line-script">command-line script</a>.
    <p>For more information about this resource, please see the GitHub readme at
      <a href="https://github.com/ewels/AWS-iGenomes">https://github.com/ewels/AWS-iGenomes</a>.</p>

    <hr>

    <h2><a name="sync-command-builder">Sync command builder</a></h2>
    <p>Use the dropdown boxes below to build an AWS S3 sync command for your reference.</p>

    <form>
      <div class="row">
        <div class="form-group col-sm-3">
          <label for="genome">Genome</label>
          <select class="form-control" id="genome">
            <option value="">[ select ]</option>
            <optgroup label="Common Genomes">
              <option value="Homo_sapiens">Homo sapiens</option>
              <option value="Mus_musculus">Mus Musculus</option>
              <option value="Saccharomyces_cerevisiae">Saccharomyces cerevisiae</option>
            </optgroup>
            <optgroup label="All Genomes" id="all_genomes"></optgroup>
          </select>
        </div>
        <div class="form-group col-sm-3">
          <label for="source">Source</label>
          <select class="form-control" id="source" disabled>
            <option value="">Please select a genome</option>
          </select>
        </div>
        <div class="form-group col-sm-3">
          <label for="build">Build</label>
          <select class="form-control" id="build" disabled>
            <option value="">Please select a source</option>
          </select>
        </div>
        <div class="form-group col-sm-3">
          <label for="type">Type</label>
          <select class="form-control" id="type" disabled>
            <option value="">Please select a build</option>
          </select>
        </div>
      </div>
      <div class="row">
        <div class="form-group col-sm-3">
          <label for="region">Region</label>
          <small class="text-muted">Currently just <code>eu-west-1</code></small>
          <select class="form-control" id="region" disabled>
            <option value="eu-west-1">eu-west-1</option>
          </select>
        </div>
        <div class="form-group col-sm-9">
          <label for="local">Local directory</label>
          <small class="text-muted">Available variables:
            <code class="local_code">{genome}</code>,
            <code class="local_code">{source}</code>,
            <code class="local_code">{build}</code>,
            <code class="local_code">{type_path}</code>
          </small>
          <input type="text" class="form-control" id="local" value="./references/{genome}/{source}/{build}/{type_path}" style="font-family:Menlo,Monaco,Consolas,'Liberation Mono','Courier New',monospace;">
        </div>
      </div>
    </form>

    <h5>Bucket URL</h5>
    <div class="prewrap" id="bucket_url_wrapper">
      <button class="btn" data-clipboard-target="#bucket_url" disabled>Copy</button>
      <pre><code id="bucket_url"><span class="text-muted">waiting for selection&hellip;</span></code></pre>
    </div>

    <h5>Sync command</h5>
    <div class="prewrap" id="sync_command_wrapper">
      <button class="btn" data-clipboard-target="#sync_command" disabled>Copy</button>
      <pre><code id="sync_command"><span class="text-muted">waiting for selection&hellip;</span></code></pre>
    </div>

    <div id="manifest_accordion" role="tablist" aria-multiselectable="true">
      <div class="card">
        <div class="card-header" role="tab" id="manifest_heading">
          <h6 class="mb-0">
            <a data-toggle="collapse" data-parent="#manifest_accordion" href="#manifest_dropdown" aria-expanded="true" aria-controls="manifest_dropdown">
              File list for selected download
            </a>
          </h6>
        </div>
        <div id="manifest_dropdown" class="collapse" role="tabpanel" aria-labelledby="manifest_heading">
          <pre class="card-block mb-0" id="file_manifest">Loading file manifest..</pre>
        </div>
      </div>
    </div>

    <hr>

    <h2><a name="command-line-script">Command-line Script</a></h2>
    <p>You can use a bash script we've written which can be used to easily obtain reference genomes
      where you need them. This builds the required command using either command-line prompts or
      flags, then runs the sync command for you.</p>
    <p>Note that you need to have the
      <a href="http://docs.aws.amazon.com/cli/latest/userguide/installing.html" target="_blank">aws command line tools</a>
      installed and set up with proper authentication.</p>

    <h5 class="mt-4">Quick usage</h5>
    <p>Paste this command in your terminal to download and run the script directly.</p>
    <div class="prewrap">
      <button class="btn btn-warning" data-clipboard-target="#aws_script_call">Copy</button>
      <pre><code id="aws_script_call">curl -fsSL https://ewels.github.io/AWS-iGenomes/aws-igenomes.sh | bash</code></pre>
    </div>

    <h5 class="mt-4">Advanced usage</h5>
    <p>You can save the script locally and run it with command-line flags as follows:</p>
    <div class="prewrap">
      <button class="btn btn-warning" data-clipboard-target="#aws_script_download">Copy</button>
      <pre><code id="aws_script_download">curl -fsSL https://ewels.github.io/AWS-iGenomes/aws-igenomes.sh > aws-igenomes.sh</code></pre>
    </div>
    <pre><code>bash aws-igenomes.sh --help

<span class="text-muted">
USAGE: aws-igenomes.sh
 		[-g &lt;genome name&gt;]
 		[-s &lt;source name&gt;]
 		[-b &lt;build name&gt;]
 		[-t &lt;reference type&gt;]
 		[-o &lt;output directory&gt;]
 		[-d (dry run, no downloads)]
 		[-q (quiet mode, non-interactive)]
 		[-h (usage help)]

All command line flags are optional. If not specified
and not running in quiet-mode, the script will prompt
for input and show available options.

Please note that this script requires the AWS command
line tools to be installed and configured for authenticated
access.</span></code></pre>

    <p><strong>Script source:</strong> <a href="https://github.com/ewels/AWS-iGenomes/blob/master/aws-igenomes.sh">https://github.com/ewels/AWS-iGenomes/blob/master/aws-igenomes.sh</a></p>
    <p><strong>Script URL:</strong> <a href="aws-igenomes.sh">https://ewels.github.io/AWS-iGenomes/aws-igenomes.sh</a></p>

    <p>If you have any questions or feature requests, please
      <a href="https://github.com/ewels/AWS-iGenomes/issues">create an issue</a>.</p>

  </div>

  <div class="footer">
    <div class="container">
      This resource was developed by <a href="http://phil.ewels.co.uk">Phil Ewels</a>, a bioinformatician
      at the <a href="https://ngisweden.scilifelab.se/">National Genomics Infrastructure</a> in Stockholm,
      Sweden. It is based on the <a href="https://support.illumina.com/sequencing/sequencing_software/igenome.html">illumina iGenomes</a>
      which has been added to with the help of <a href="https://www.uppmax.uu.se/">UPPMAX</a>.
      S3 hosting fees are kindly covered by a grant supplied by Amazon.
      <div class="row mt-4">
        <div class="col-md-4 text-center"><a href="https://aws.amazon.com/grants/"><img title="AWS Programs for Research and Education" src="images/AWS_logo.png"></a></div>
        <div class="col-md-4 text-center"><a href="http://www.scilifelab.se/"><img title="SciLifeLab" src="images/SciLifeLab_logo.png" class="sll-logo"></a></div>
        <div class="col-md-4 text-center"><a href="https://ngisweden.scilifelab.se/"><img title="National Genomics Infrastructure" src="images/NGI_logo.png"></a></div>
      </div>
    </div>
  </div>


  <!--  JavaScript -->
  <script src="https://code.jquery.com/jquery-3.2.1.min.js" integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/tether/1.4.0/js/tether.min.js" integrity="sha384-DztdAPBWPRXSA/3eYEEUWrWCy7G5KFbe8fFjk5JAIxUYHKkDx6Qin1DkWx51bBrb" crossorigin="anonymous"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-alpha.6/js/bootstrap.min.js" integrity="sha384-vBWWzlZJ8ea9aCX4pEW3rVHjgjt7zpkNpZk+02D9phzyeVkE+jo0ieGizqPLForn" crossorigin="anonymous"></script>
  <script src="https://cdn.jsdelivr.net/clipboard.js/1.6.1/clipboard.min.js"></script>
  <script type="text/javascript">
  var types = {
      "Arabidopsis_thaliana/Ensembl/TAIR10": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Arabidopsis_thaliana/Ensembl/TAIR9": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Arabidopsis_thaliana/NCBI/build9.1": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Arabidopsis_thaliana/NCBI/TAIR10": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bacillus_cereus_ATCC_10987/NCBI/2004-02-13": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Bacillus_subtilis_168/Ensembl/EB2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Bos_taurus/Ensembl/Btau_4.0": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bos_taurus/Ensembl/UMD3.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Bos_taurus/NCBI/Btau_4.2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bos_taurus/NCBI/Btau_4.6.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bos_taurus/NCBI/UMD_3.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bos_taurus/NCBI/UMD_3.1.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bos_taurus/UCSC/bosTau4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bos_taurus/UCSC/bosTau6": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bos_taurus/UCSC/bosTau7": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Bos_taurus/UCSC/bosTau8": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Caenorhabditis_elegans/Ensembl/WBcel215": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Caenorhabditis_elegans/Ensembl/WBcel235": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Caenorhabditis_elegans/Ensembl/WS210": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Caenorhabditis_elegans/Ensembl/WS220": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Caenorhabditis_elegans/NCBI/WS190": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Caenorhabditis_elegans/NCBI/WS195": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Caenorhabditis_elegans/UCSC/ce10": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Caenorhabditis_elegans/UCSC/ce6": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Canis_familiaris/Ensembl/BROADD2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Canis_familiaris/Ensembl/CanFam3.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Canis_familiaris/NCBI/build2.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Canis_familiaris/NCBI/build3.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Canis_familiaris/UCSC/canFam2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Canis_familiaris/UCSC/canFam3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Danio_rerio/Ensembl/GRCz10": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Danio_rerio/Ensembl/Zv9": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Danio_rerio/NCBI/GRCz10": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Danio_rerio/NCBI/Zv9": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Danio_rerio/UCSC/danRer10": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Danio_rerio/UCSC/danRer7": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/Ensembl/BDGP5": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/Ensembl/BDGP5.25": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/Ensembl/BDGP6": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/NCBI/build4.1": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/NCBI/build5": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/NCBI/build5.3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/NCBI/build5.41": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/UCSC/dm3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Drosophila_melanogaster/UCSC/dm6": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Enterobacteriophage_lambda/NCBI/1993-04-28": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Equus_caballus/Ensembl/EquCab2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Equus_caballus/NCBI/EquCab2.0": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Equus_caballus/UCSC/equCab2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Escherichia_coli_K_12_DH10B/Ensembl/EB1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Escherichia_coli_K_12_DH10B/NCBI/2008-03-17": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Escherichia_coli_K_12_MG1655/NCBI/2001-10-15": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Gallus_gallus/Ensembl/Galgal4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Gallus_gallus/Ensembl/WASHUC2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Gallus_gallus/NCBI/build2.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Gallus_gallus/NCBI/build3.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Gallus_gallus/UCSC/galGal3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Gallus_gallus/UCSC/galGal4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Glycine_max/Ensembl/Gm01": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Homo_sapiens/Ensembl/GRCh37": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Homo_sapiens/GATK/b37": ["gatk"],
      "Homo_sapiens/GATK/hg19": ["gatk"],
      "Homo_sapiens/GATK/hg38": ["gatk"],
      "Homo_sapiens/NCBI/build36.3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Homo_sapiens/NCBI/build37.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Homo_sapiens/NCBI/build37.2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Homo_sapiens/NCBI/GRCh38": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Homo_sapiens/NCBI/GRCh38Decoy": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Homo_sapiens/UCSC/hg18": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Homo_sapiens/UCSC/hg19": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Homo_sapiens/UCSC/hg38": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Macaca_mulatta/Ensembl/Mmul_1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Mus_musculus/Ensembl/GRCm38": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Mus_musculus/Ensembl/NCBIM37": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Mus_musculus/NCBI/build37.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Mus_musculus/NCBI/build37.2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Mus_musculus/NCBI/GRCm38": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Mus_musculus/UCSC/mm10": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Mus_musculus/UCSC/mm9": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Mycobacterium_tuberculosis_H37RV/Ensembl/H37Rv.EB1": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Mycobacterium_tuberculosis_H37RV/NCBI/2001-09-07": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Oryza_sativa_japonica/Ensembl/IRGSP-1.0": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Oryza_sativa_japonica/Ensembl/MSU6": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Pan_troglodytes/Ensembl/CHIMP2.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Pan_troglodytes/Ensembl/CHIMP2.1.4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Pan_troglodytes/NCBI/build2.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Pan_troglodytes/NCBI/build3.1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Pan_troglodytes/UCSC/panTro2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Pan_troglodytes/UCSC/panTro3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Pan_troglodytes/UCSC/panTro4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "PhiX/Illumina/RTA": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "PhiX/NCBI/1993-04-28": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Pseudomonas_aeruginosa_PAO1/NCBI/2000-09-13": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Rattus_norvegicus/Ensembl/RGSC3.4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Rattus_norvegicus/Ensembl/Rnor_5.0": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Rattus_norvegicus/Ensembl/Rnor_6.0": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Rattus_norvegicus/NCBI/RGSC_v3.4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Rattus_norvegicus/NCBI/Rnor_5.0": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Rattus_norvegicus/NCBI/Rnor_6.0": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Rattus_norvegicus/UCSC/rn4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Rattus_norvegicus/UCSC/rn5": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Rattus_norvegicus/UCSC/rn6": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Rhodobacter_sphaeroides_2.4.1/NCBI/2005-10-07": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Saccharomyces_cerevisiae/Ensembl/EF2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Saccharomyces_cerevisiae/Ensembl/EF3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Saccharomyces_cerevisiae/Ensembl/EF4": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Saccharomyces_cerevisiae/Ensembl/R64-1-1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Saccharomyces_cerevisiae/NCBI/build2.1": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Saccharomyces_cerevisiae/NCBI/build3.1": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Saccharomyces_cerevisiae/UCSC/sacCer2": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Saccharomyces_cerevisiae/UCSC/sacCer3": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Schizosaccharomyces_pombe/Ensembl/EF1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Schizosaccharomyces_pombe/Ensembl/EF2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Sorangium_cellulosum_So_ce_56/NCBI/2007-11-27": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Sorghum_bicolor/Ensembl/Sbi1": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Staphylococcus_aureus_NCTC_8325/NCBI/2006-02-13": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq"],
      "Sus_scrofa/Ensembl/Sscrofa10.2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Sus_scrofa/Ensembl/Sscrofa9": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna", "variation"],
      "Sus_scrofa/NCBI/Sscrofa10": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Sus_scrofa/NCBI/Sscrofa10.2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Sus_scrofa/NCBI/Sscrofa9.2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Sus_scrofa/UCSC/susScr2": ["bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Sus_scrofa/UCSC/susScr3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Zea_mays/Ensembl/AGPv2": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"],
      "Zea_mays/Ensembl/AGPv3": ["gtf", "bed12", "bismark", "bowtie", "bowtie2", "bwa", "star", "fasta", "chromosomes", "abundantseq", "smrna"]
  };
  var type_nicenames = {
      gtf: "GTF",
      bed12: "BED",
      bismark: "Bismark",
      bowtie: "Bowtie 1",
      bowtie2: "Bowtie 2",
      bwa: "BWA",
      star: "STAR",
      fasta: "FASTA",
      chromosomes: "FASTA (Chromosomes)",
      abundantseq: "Abundant Sequences",
      smrna: "smRNA",
      variation: "Variation",
      gatk: "GATK"
  };
  var type_suffixes = {
      gtf: "Annotation/Genes/", // appended to command: genes.gtf
      bed12: "Annotation/Genes/", // appended to command: genes.bed
      bismark: "Sequence/BismarkIndex/",
      bowtie: "Sequence/BowtieIndex/",
      bowtie2: "Sequence/Bowtie2Index/",
      bwa: "Sequence/BWAIndex/",
      star: "Sequence/STARIndex/",
      fasta: "Sequence/WholeGenomeFasta/",
      chromosomes: "Sequence/Chromosomes/",
      abundantseq: "Sequence/AbundantSequences/",
      smrna: "Annotation/SmallRNA/",
      variation: "Annotation/Variation/",
      gatk: ""
  };
  var manifest = [];
  $(function(){
    // Initiate clipboard.js
    new Clipboard('.prewrap > .btn');

    // Get the full file list
    try {
      $.get('ngi-igenomes_file_manifest.txt', function(data){
        try {
          manifest = data.match(/[^\r\n]+/g);
          $('#file_manifest').html('waiting for selection..');
        } catch(e){
          $('#file_manifest').html('Error - could not load file manifest.').attr('id', 'file_manifest_error');
          console.error(e);
        }
      }).fail(function() {
        $('#file_manifest').html('Error - could not load file manifest.').attr('id', 'file_manifest_error');
      });
    } catch(e){
      $('#file_manifest').html('Error - could not load file manifest.').attr('id', 'file_manifest_error');
      console.error(e);
    }

    // Build our variables
    var paths = Object.keys(types);
    var genomes = [];
    var sources = {};
    var builds = {};
    for(i=0; i<paths.length; i++){
      var s = paths[i].split('/');
      if(genomes.indexOf(s[0]) < 0){
        genomes.push(s[0]);
      }
      if(sources[s[0]] == undefined){
        sources[s[0]] = [];
      }
      if(sources[s[0]].indexOf(s[1]) < 0){
        sources[s[0]].push(s[1]);
      }
      if(builds[s[0]+'/'+s[1]] == undefined){
        builds[s[0]+'/'+s[1]] = [];
      }
      builds[s[0]+'/'+s[1]].push(s[2]);
    }
    // Populate first dropdown
    for(i=0; i<genomes.length; i++){
      $('#all_genomes').append($('<option>', {
        value: genomes[i],
        text: genomes[i].replace('_',' ')
      }));
    }
    // LISTENERS!
    $('#genome').change(function(e){
      e.preventDefault();
      // Reset downstream dropdown
      $('#build').html('<option value="">Please select a source</option>').attr('disabled', true);
      $('#type').html('<option value="">Please select a build</option>').attr('disabled', true);
      $('#bucket_url, #sync_command').html('<span class="text-muted">waiting for selection…</span>');
      // Set up source dropdown
      var g = $(this).val();
      if(g == ''){
        $('#source').html('<option value="">Please select a genome</option>').attr('disabled', true);
      } else {
        $('#source').html('').attr('disabled', false);
        if(sources[g].length > 1){
          $('#source').html('<option value="">[ select ]</option>');
        }
        for(i=0; i<sources[g].length; i++){
          $('#source').append($('<option>', {
            value: sources[g][i],
            text: sources[g][i]
          }));
        }
        $('#source').trigger('change');
        $('#build').trigger('change');
        $('#type').trigger('change');
      }
    });
    // Source dropdown
    $('#source').change(function(e){
      e.preventDefault();
      // Reset downstream dropdown
      $('#type').html('<option value="">Please select a build</option>').attr('disabled', true);
      $('#bucket_url, #sync_command').html('<span class="text-muted">waiting for selection…</span>');
      // Set up build dropdown
      var g = $('#genome').val();
      var s = $(this).val();
      var k = g+'/'+s;
      if(s == ''){
        $('#build').html('<option value="">Please select a source</option>').attr('disabled', true);
      } else {
        $('#build').html('').attr('disabled', false);
        if(builds[k].length > 1){
          $('#build').html('<option value="">[ select ]</option>');
        }
        for(i=0; i<builds[k].length; i++){
          $('#build').append($('<option>', {
            value: builds[k][i],
            text: builds[k][i]
          }));
        }
        $('#build').trigger('change');
        $('#type').trigger('change');
      }
    });
    // Build dropdown
    $('#build').change(function(e){
      e.preventDefault();
      // Set up type dropdown
      var g = $('#genome').val();
      var s = $('#source').val();
      var b = $(this).val();
      var k = g+'/'+s+'/'+b;
      if(b == ''){
        $('#type').html('<option value="">Please select a source</option>').attr('disabled', true);
        $('#bucket_url, #sync_command').html('<span class="text-muted">waiting for selection…</span>');
      } else {
        $('#type').html('').attr('disabled', false);
        if(types[k].length > 1){
          $('#type').html('<option value="">[ select ]</option>');
        }
        for(i=0; i<types[k].length; i++){
          $('#type').append($('<option>', {
            value: types[k][i],
            text: type_nicenames[types[k][i]]
          }));
        }
        $('#type').trigger('change');
      }
    });
    // Type dropdown
    $('#type, #region, #local').change(function(e){ update_cmd(); });
    $('#local').keyup(function(e){ update_cmd(); });
    function update_cmd(){
      var g = $('#genome').val();
      var s = $('#source').val();
      var b = $('#build').val();
      var t = $('#type').val();
      var region = $('#region').val();
      var local = $('#local').val();
      if(g == '' || s == '' || b == '' || t == '' || region == '' || local == ''){
        $('#bucket_url, #sync_command, #file_manifest').html('<span class="text-muted">waiting for selection…</span>');
        $('#bucket_url_wrapper button, #sync_command_wrapper button')
          .removeClass('btn-warning')
          .attr('disabled', true);
      } else {
        // Build path and sync command
        var path = g+'/'+s+'/'+b+'/'+type_suffixes[t];
        local = local.replace('{genome}',g).replace('{source}',s).replace('{build}',b).replace('{type_path}',type_suffixes[t]);
        var s3path = 's3://ngi-igenomes/igenomes/'+path;
        var s3cmd = 'aws s3 --no-sign-request --region '+region+' sync '+s3path+' '+local;
        if(t == 'gtf'){
          s3path += 'genes.gtf';
          s3cmd += ' --exclude "*" --include "genes.gtf"';
        }
        if(t == 'bed12'){
          s3path += 'genes.bed';
          s3cmd += ' --exclude "*" --include "genes.bed"';
        }
        $('#bucket_url').html(s3path);
        $('#sync_command').html(s3cmd);
        $('#bucket_url_wrapper button, #sync_command_wrapper button')
          .addClass('btn-warning')
          .attr('disabled', false);

        // Get the file manifest
        if(manifest.length > 0){
          var files = '';
          for(i=0; i<manifest.length; i++){
            if(manifest[i].substr(0, s3path.length) == s3path){
              files += manifest[i] + "\n";
            }
          }
          $('#file_manifest').html(files);
        }
      }
    }

    // Code example clicks
    $('.local_code').click(function(e){
      var ns = $('#local').val() + $(this).text();
      $('#local').val(ns);
    });

  });
  </script>
</body>
</html>
